File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/XCertificateInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import kotlinx.serialization.Serializable
import org.bouncycastle.cert.X509AttributeCertificateHolder
import org.bouncycastle.cert.X509CertificateHolder

@Serializable
sealed class XCertificateInfo : EncodableInfo {
  companion object {
    fun getInstance(obj: Any): XCertificateInfo =
      when (obj) {
        is X509CertificateHolder -> X509CertificateInfo.getInstance(obj)
        is X509AttributeCertificateHolder -> X509AttributeCertificateInfo.getInstance(obj)
        else -> throw IllegalArgumentException("Unexpected certificate type")
      }
  }
}

File: net/JetBrains.Serialization/src/Dataholders/XCertificateInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;
using JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public abstract class XCertificateInfo : IEncodableInfo
{
  public static XCertificateInfo GetInstance(Asn1Encodable obj)
  {
    switch (obj)
    {
      case TbsCertificateStructure tbs:
        return X509CertificateInfo.GetInstance(tbs);
      case AttributeCertificate attributeCertificate:
        return X509AttributeCertificateInfo.GetInstance(attributeCertificate);
      default:
        throw new ArgumentException(@"Unexpected certificate type", nameof(obj));
    }
  }

  public abstract Asn1Encodable ToPrimitive();
}